jlm/llvm/opt/alias-analyses/Andersen.cpp

```diff
@@ -0,0 +1,514 @@
+   */
+  bool
+  AddToPointsToSet(PointerObjectIndex pointer, PointerObjectIndex pointee)
+  {
+    JLM_ASSERT(pointer < NodeCount());
+    JLM_ASSERT(pointee < NodeCount());
+    JLM_ASSERT(GetNode(pointee).GetKind() != PointerObjectKind::Register);
+
+    auto sizeBefore = PointsToSets_[pointer].size();
+    PointsToSets_[pointer].insert(pointee);
+
+    return PointsToSets_[pointer].size() != sizeBefore;
+  }
+
+  /**
+   * Makes P(superset) a superset of P(subset) by adding any elements in the set difference
+   * @return true if P(superset) was changed by this
+   */
+  bool
+  MakePointsToSetSuperset(PointerObjectIndex superset, PointerObjectIndex subset)
+  {
+    auto& P_super = PointsToSets_[superset];
+    auto& P_sub = PointsToSets_[subset];
+
+    auto sizeBefore = P_super.size();
+    P_super.insert(P_sub.begin(), P_sub.end());
+
+    return P_super.size() != sizeBefore;
+  }
+
+  PointerObjectIndex
+  CreateRegisterNode(jlm::rvsdg::output *rvsdgOutput)
+  {
+    JLM_ASSERT(RegisterMap_.count(rvsdgOutput) == 0);
+    return RegisterMap_[rvsdgOutput] = AddNode(PointerObjectKind::Register);
+  }
+
+  void
+  MapRegisterToExistingNode(jlm::rvsdg::output *rvsdgOutput, PointerObjectIndex node)
+  {
+    JLM_ASSERT(RegisterMap_.count(rvsdgOutput) == 0);
+    JLM_ASSERT(node < NodeCount());
+    RegisterMap_[rvsdgOutput] = node;
+  }
+
+  PointerObjectIndex
+  CreateAllocaMemoryObject(jlm::rvsdg::node *allocaNode)
+  {
+    JLM_ASSERT(AllocaMap_.count(allocaNode) == 0);
+    return AllocaMap_[allocaNode] = AddNode(PointerObjectKind::AllocaMemoryObject);
+  }
+
+  PointerObjectIndex
+  CreateMallocMemoryObject(jlm::rvsdg::node *mallocNode)
+  {
+    JLM_ASSERT(MallocMap_.count(mallocNode) == 0);
+    return MallocMap_[mallocNode] = AddNode(PointerObjectKind::MallocMemoryObject);
+  }
+
+  PointerObjectIndex
+  CreateGlobalMemoryObject(jlm::llvm::delta::node *deltaNode)
+  {
+    JLM_ASSERT(GlobalMap_.count(deltaNode) == 0);
+    return GlobalMap_[deltaNode] = AddNode(PointerObjectKind::GlobalMemoryObject);
+  }
+
+  PointerObjectIndex
+  CreateFunction(jlm::llvm::lambda::node *lambdaNode)
+  {
+    JLM_ASSERT(FunctionMap_.count(lambdaNode) == 0);
+    return FunctionMap_[lambdaNode] = AddNode(PointerObjectKind::Function);
+  }
+
+  PointerObjectIndex
+  CreateImportMemoryObject(jlm::rvsdg::argument *importNode)
+  {
+    JLM_ASSERT(ImportMap_.count(importNode) == 0);
+    return ImportMap_[importNode] = AddNode(PointerObjectKind::ImportMemoryObject);
+  }
+
+  /**
+   * Converts the PointerObjectSet nodes into PointsToGraph nodes,
+   * and points-to-graph set memberships into edges.
+   *
+   * Note that registers sharing PointerObject, become separate PointsToGraph nodes.
+   *
+   * The *escaped* node is not included.
+   * Instead implicit edges through escaped+external, are added as explicit edges.
+   * @return
+   */
+  std::unique_ptr<PointsToGraph>
+  ConstructPointsToGraph()
```

Comment:
The points-to graph construction method as part of the `PointerObjectSet` class feels "wrong". Should it not rather be a static private method of the Andersen class and you hand the set in as parameter? This would require that you define access methods to the private attributes though....